proxy/src/control/cache.rs

```diff
@@ -70,61 +70,82 @@ where
     pub fn update_union<I, F>(&mut self, iter: I, on_change: &mut F)
     where
         I: Iterator<Item = (K, V)>,
-        F: FnMut((K, V), CacheChange),
+        F: for<'value> FnMut(CacheChange<'value, K, V>),
     {
-        fn update_inner<K, V, I, F>(
-            inner: &mut IndexMap<K, V>,
-            iter: I,
+        fn update_inner<K, V, F>(
+            inner: & mut IndexMap<K, V>,
+            key: K,
+            new_value: V,
             on_change: &mut F
         )
         where
-            K: Eq + Hash + Copy + Clone,
-            V: PartialEq + Clone,
-            I: Iterator<Item = (K, V)>,
-            F: FnMut((K, V), CacheChange),
+            K: Eq + Hash + Copy,
+            V: PartialEq,
+            F: for<'value> FnMut(CacheChange<'value, K, V>),
         {
-            for (key, value) in iter {
-                match inner.insert(key, value.clone()) {
-                    // If the returned value is equal to the inserted value,
-                    // then the inserted key was already present in the map
-                    // and the value is unchanged. Do nothing.
-                    Some(ref old_value) if old_value == &value => {},
-                    // If `insert` returns `Some` with a different value than
-                    // the one we inserted, then we changed the value for that
-                    // key.
-                    Some(_) =>
-                        on_change((key, value), CacheChange::Modification),
+            if inner.get(&key)
```

Comment:
You call `inner.get(&key)` here because `inner.insert(key, new_value)` consumes `new_value`, right? However, isn't this cure worse than the problem? In particular, `inner.get(&key)` would likely be as slow or slower than cloning `new_value`, right?

I think instead we could instead do something like this, with one lookup:

```rust
inner.entry(&key)
    .and_modify(|old| {
        if *old != new {
            *old = new;
            on_change(....)
        }
    })
    .or_insert(...);
```

Or, the equivalent using pattern matching on the result of `inner.entry(&key)`.

Do you think that using the entry API solves this problem and/or the other related issues?

Reply:
@briansmith that does, in fact, solve this problem, thanks for the suggestion.

As a side note, you'll notice that I used pattern matching on the entry rather than `.and_modify(...).or_insert(...)` as you suggested. This is actually because `IndexMap`'s entry API lacks an `and_modify` method --- only `std::collections`'s maps have that method.